sklearn/metrics/pairwise.py

```diff
@@ -829,6 +834,172 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def _split_categorical_numerical(X, categorical_features):
+    # the following bit is done before check_pairwise_array to avoid converting
+    # numerical data to object dtype. First we split the data into categorical
+    # and numerical, then we do check_array
+
+    if X is None:
+        return None, None
+
+    # TODO: this should be more like check_array(..., accept_pandas=True)
+    if not hasattr(X, "shape"):
+        X = check_array(X, dtype=np.object, force_all_finite=False)
+
+    if callable(categorical_features):
+        cols = categorical_features(X)
+    else:
+        cols = categorical_features
+    if cols is None:
+        cols = []
+
+    col_idx = _get_column_indices(X, cols)
+    X_cat = _safe_indexing(X, col_idx, axis=1)
```

Comment:
Do we care that this is (I think) always a copying operation even if col_idx is empty?